text/2338-type-alias-enum-variants.md

```diff
@@ -0,0 +1,120 @@
+- Feature Name: `type_alias_enum_variants`
+- Start Date: 2018-02-15
+- RFC PR: [rust-lang/rfcs#2338](https://github.com/rust-lang/rfcs/pull/2338)
+- Rust Issue: [rust-lang/rust#49683](https://github.com/rust-lang/rust/issues/49683)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes to allow access to enum variants through type aliases. This
+enables better abstraction/information hiding by encapsulating enums in aliases
+without having to create another enum type and requiring the conversion from
+and into the "alias" enum.
+
+# Motivation
+[motivation]: #motivation
+
+While type aliases provide a useful means of encapsulating a type definition in
+order to hide implementation details or provide a more ergonomic API, the
+substitution principle currently falls down in the face of enum variants. It's
+reasonable to expect that a type alias can fully replace the original type
+specification, and so the lack of working support for aliased enum variants
+represents an ergonomic gap in the language/type system. This can be useful in
+exposing an interface from a dependency to library users while "hiding" the exact
+implementation details. There's at least some evidence that people have asked
+about this capability before.
+
+Since `Self` also works as an alias, this should also enable the use of `Self`
+in more places.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In general, the simple explanation here is that type aliases can be used in
+more places where you currently have to go through the original type definition,
+as it relates to enum variants. As much as possible, enum variants should work
+as if the original type was specified rather than the alias. This should make
+type aliases easier to learn than before, because there are fewer exceptions
+to their applicability.
+
+```rust
+enum Foo {
+    Bar(i32),
+    Baz { i: i32 },
+}
+
+type Alias = Foo;
+
+fn main() {
+    let t = Alias::Bar(0);
+    let t = Alias::Baz { i: 0 };
+    match t {
+        Alias::Bar(_i) => {}
+        Alias::Baz { i: _i } => {}
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+If a path refers into an alias, the behavior for enum variants should be as
+if the alias was substituted with the original type. Here are some examples of
+the new behavior in edge cases:
+
+```rust
+type Alias<T> = Option<T>;
+
+mod foo {
+    pub use Alias::Some;
```

Comment:
I'm inclined to think this shouldn't be allowed, or at least it shouldn't be part of the initial implementation. After all, even
```rust
pub use Option::Some;
```
does not work right now, although
```rust
pub use ::std::option::Option::Some;
```
does, however.
For me, the semantics are unclear, and I'd be inclined not to accept it by default.

Anyway, would appreciate the thoughts of @rust-lang/lang on this matter.

Reply:
Yes, this can't work right now.
`Option::Some` is module-relative (enums and traits are kind of modules from name resolution's point of view), so they can be used in imports.
`Alias::Some` is type-relative, and we need type checking to determine what `Alias` actually refers to, so it cannot be used in imports due to phase separation.